File: C++/PluginSystemVersion/PluginSystemVersion.cpp
/* Copyright (C) 2017 Rainmeter Project Developers
 *
 * This Source Code Form is subject to the terms of the GNU General Public
 * License; either version 2 of the License, or (at your option) any later
 * version. If a copy of the GPL was not distributed with this file, You can
 * obtain one at <https://www.gnu.org/licenses/gpl-2.0.html>. */

#include <Windows.h>
#include <cstdio>
#include <string>
#include "../../API/RainmeterAPI.h"

// Overview: This example demonstrates the basic concept of Rainmeter C++ plugins.

// Sample skin:
/*
	[Rainmeter]
	Update=1000
	DynamicWindowSize=1
	BackgroundMode=2
	SolidColor=255,255,255

	[mString]
	Measure=Plugin
	Plugin=SystemVersion
	Type=String

	[mMajor]
	Measure=Plugin
	Plugin=SystemVersion
	Type=Major

	[mMinor]
	Measure=Plugin
	Plugin=SystemVersion
	Type=Minor

	[mNumber]
	Measure=Plugin
	Plugin=SystemVersion
	Type=Number

	[Text1]
	Meter=String
	MeasureName=mString
	MeasureName2=mMajor
	MeasureName3=mMinor
	MeasureName4=mNumber
	Text="String: %1#CRLF#Major: %2#CRLF#Minor: %3#CRLF#Number: %4#CRLF#"

	[Text2]
	Meter=String
	MeasureName=mString
	MeasureName2=mMajor
	MeasureName3=mMinor
	MeasureName4=mNumber
	NumOfDecimals=1
	Y=5R
	Text="String: %1#CRLF#Major: %2#CRLF#Minor: %3#CRLF#Number: %4#CRLF#"
*/

enum MeasureType
{
	MEASURE_MAJOR,
	MEASURE_MINOR,
	MEASURE_NUMBER,
	MEASURE_STRING
};

struct Measure
{
	MeasureType type;
	std::wstring strValue;

	Measure() :
		type(MEASURE_MAJOR),
		strValue() {}
};

PLUGIN_EXPORT void Initialize(void** data, void* rm)
{
	Measure* measure = new Measure;
	*data = measure;
}

PLUGIN_EXPORT void Reload(void* data, void* rm, double* maxValue)
{
	Measure* measure = (Measure*)data;

	LPCWSTR value = RmReadString(rm, L"Type", L"");
	if (_wcsicmp(value, L"Major") == 0)
	{
		measure->type = MEASURE_MAJOR;
	}
	else if (_wcsicmp(value, L"Minor") == 0)
	{
		measure->type = MEASURE_MINOR;
	}
	else if (_wcsicmp(value, L"Number") == 0)
	{
		measure->type = MEASURE_NUMBER;
	}
	else if (_wcsicmp(value, L"String") == 0)
	{
		measure->type = MEASURE_STRING;
	}
	else
	{
		RmLog(rm, LOG_ERROR, L"Invalid \"Type\"");
	}
}

PLUGIN_EXPORT double Update(void* data)
{
	Measure* measure = (Measure*)data;

	OSVERSIONINFOEX osvi = {sizeof(OSVERSIONINFOEX)};
	if (!GetVersionEx((OSVERSIONINFO*)&osvi))
	{
		return 0.0;
	}

	switch (measure->type)
	{
	case MEASURE_MINOR:
		return (double)osvi.dwMinorVersion;

	case MEASURE_MAJOR:
		return (double)osvi.dwMajorVersion;

	case MEASURE_NUMBER:
		return (double)osvi.dwMajorVersion + ((double)osvi.dwMinorVersion / 10.0);

	case MEASURE_STRING:
		{
			WCHAR buffer[128];
			_snwprintf_s(buffer, _TRUNCATE, L"%i.%i (Build %i)",
				(int)osvi.dwMajorVersion, (int)osvi.dwMinorVersion, (int)osvi.dwBuildNumber);
			measure->strValue = buffer;
		}
		break;
	}

	// MEASURE_STRING is not a number and therefore will be returned in GetString. However it is
	//  recommended that you do any processing here in the Update function rather than the GetString
	//  function since GetString can be called multiple times during the update cycle. Store any
	//  string values in your Measure structure and return it in GetString.

	return 0.0;
}

PLUGIN_EXPORT LPCWSTR GetString(void* data)
{
	Measure* measure = (Measure*)data;

	// Although GetVersionEx is a rather inexpensive operation, it is recommended to
	// do any processing in the Update function. See the comments above.
	if (!measure->strValue.empty())
	{
		switch (measure->type)
		{
		case MEASURE_STRING:
			return measure->strValue.c_str();
		}
	}

	// MEASURE_MAJOR, MEASURE_MINOR, and MEASURE_NUMBER are numbers. Therefore,
	// |nullptr| is returned here for them. This is to inform Rainmeter that it can
	// treat those types as numbers.

	return nullptr;
}

PLUGIN_EXPORT void Finalize(void* data)
{
	Measure* measure = (Measure*)data;
	delete measure;
}

File: C#/PluginSystemVersion/PluginSystemVersion.cs
/*
  Copyright (C) 2014 Birunthan Mohanathas

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using Rainmeter;

// Overview: This example demonstrates the basic concept of Rainmeter C# plugins.

// Sample skin:
/*
    [Rainmeter]
    Update=1000
    BackgroundMode=2
    SolidColor=000000

    [mString]
    Measure=Plugin
    Plugin=SystemVersion.dll
    Type=String

    [mMajor]
    Measure=Plugin
    Plugin=SystemVersion.dll
    Type=Major

    [mMinor]
    Measure=Plugin
    Plugin=SystemVersion.dll
    Type=Minor

    [mNumber]
    Measure=Plugin
    Plugin=SystemVersion.dll
    Type=Number

    [Text1]
    Meter=STRING
    MeasureName=mString
    MeasureName2=mMajor
    MeasureName3=mMinor
    MeasureName4=mNumber
    X=5
    Y=5
    W=300
    H=70
    FontColor=FFFFFF
    Text="String: %1#CRLF#Major: %2#CRLF#Minor: %3#CRLF#Number: %4#CRLF#"

    [Text2]
    Meter=STRING
    MeasureName=mString
    MeasureName2=mMajor
    MeasureName3=mMinor
    MeasureName4=mNumber
    NumOfDecimals=1
    X=5
    Y=5R
    W=300
    H=70
    FontColor=FFFFFF
    Text="String: %1#CRLF#Major: %2#CRLF#Minor: %3#CRLF#Number: %4#CRLF#"
*/

namespace PluginSystemVersion
{
    internal class Measure
    {
        enum MeasureType
        {
            Major,
            Minor,
            Number,
            String
        }

        private MeasureType Type = MeasureType.Major;

        internal Measure()
        {
        }

        internal void Reload(Rainmeter.API api, ref double maxValue)
        {
            string type = api.ReadString("Type", "");
            switch (type.ToLowerInvariant())
            {
                case "major":
                    Type = MeasureType.Major;
                    break;

                case "minor":
                    Type = MeasureType.Minor;
                    break;

                case "number":
                    Type = MeasureType.Number;
                    break;

                case "string":
                    Type = MeasureType.String;
                    break;

                default:
                    api.Log(API.LogType.Error, "SystemVersion.dll: Type=" + type + " not valid");
                    break;
            }
        }

        internal double Update()
        {
            switch (Type)
            {
                case MeasureType.Major:
                    return (double)Environment.OSVersion.Version.Major;

                case MeasureType.Minor:
                    return (double)Environment.OSVersion.Version.Minor;

                case MeasureType.Number:
                    return (double)Environment.OSVersion.Version.Major + ((double)Environment.OSVersion.Version.Minor / 10.0);
            }

            // MeasureType.MajorMinor is not a number and and therefore will be
            // returned in GetString.

            return 0.0;
        }

        internal string GetString()
        {
            switch (Type)
            {
                case MeasureType.String:
                    return string.Format("{0}.{1} (Build {2})", Environment.OSVersion.Version.Major, Environment.OSVersion.Version.Minor, Environment.OSVersion.Version.Build);
            }

            // MeasureType.Major, MeasureType.Minor, and MeasureType.Number are
            // numbers. Therefore, null is returned here for them. This is to
            // inform Rainmeter that it can treat those types as numbers.

            return null;
        }
    }

    public static class Plugin
    {
        static IntPtr StringBuffer = IntPtr.Zero;

        [DllExport]
        public static void Initialize(ref IntPtr data, IntPtr rm)
        {
            data = GCHandle.ToIntPtr(GCHandle.Alloc(new Measure()));
        }

        [DllExport]
        public static void Finalize(IntPtr data)
        {
            GCHandle.FromIntPtr(data).Free();

            if (StringBuffer != IntPtr.Zero)
            {
                Marshal.FreeHGlobal(StringBuffer);
                StringBuffer = IntPtr.Zero;
            }
        }

        [DllExport]
        public static void Reload(IntPtr data, IntPtr rm, ref double maxValue)
        {
            Measure measure = (Measure)GCHandle.FromIntPtr(data).Target;
            measure.Reload(new Rainmeter.API(rm), ref maxValue);
        }

        [DllExport]
        public static double Update(IntPtr data)
        {
            Measure measure = (Measure)GCHandle.FromIntPtr(data).Target;
            return measure.Update();
        }

        [DllExport]
        public static IntPtr GetString(IntPtr data)
        {
            Measure measure = (Measure)GCHandle.FromIntPtr(data).Target;
            if (StringBuffer != IntPtr.Zero)
            {
                Marshal.FreeHGlobal(StringBuffer);
                StringBuffer = IntPtr.Zero;
            }

            string stringValue = measure.GetString();
            if (stringValue != null)
            {
                StringBuffer = Marshal.StringToHGlobalUni(stringValue);
            }

            return StringBuffer;
        }
    }
}
